third_party/move/move-vm/runtime/src/loader/mod.rs

```diff
@@ -1929,13 +1974,36 @@ impl Loader {
                 subst_struct_formula.scale(1);
                 subst_struct_formula
             },
+            Type::Function {
+                args,
+                results,
+                abilities: _,
+            } => {
+                let mut inner = DepthFormula::normalize(
+                    args.iter()
+                        .chain(results)
+                        .map(|arg_ty| {
+                            self.calculate_depth_of_type(
+                                arg_ty,
+                                module_store,
+                                module_storage,
+                                visited_cache,
+                            )
+                        })
+                        .collect::<PartialVMResult<Vec<_>>>()?,
+                );
+                inner.scale(1);
+                inner
+            },
         })
     }
 }
 
 // Matches the actual returned type to the expected type, binding any type args to the
 // necessary type as stored in the map. The expected type must be a concrete type (no TyParam).
 // Returns true if a successful match is made.
+// TODO: is this really needed in presence of paranoid mode? This does a deep structural
+//   comparison and is expensive.
 fn match_return_type<'a>(
```

Comment:
IIRC this function is needed for handling entry function with returned types needed by Sui, we didn't need such functionality and can probably clean up such logic later

Reply:
Yep, Gerben introduced this IIRC.